File: minimumSwaps.ts
// Source: https://www.hackerrank.com/challenges/minimum-swaps-2/problem
export const minimumSwaps = (arr: number[]): number => {
  // Instantiate swaps
  let swaps: number = 0;

  // Create mapping of index <-> value using objects
  const idxToVal: any = {};
  const valToIdx: any = {};
  for (let i = 0; i < arr.length; ++i) {
    idxToVal[i] = arr[i];
    valToIdx[arr[i] as number] = i;
  }

  // Iterate through each index
  for (let current = 0; current < arr.length; ++current) {
    // If index does not contain proper value
    if (arr[current] !== current + 1) {
      // Swap the current value with the proper one with the help of the map
      const target: number = valToIdx[current + 1];
      [arr[current], arr[target]] = [
        arr[target] as number,
        arr[current] as number,
      ];
      swaps += 1;

      // Update the map values
      [idxToVal[current], idxToVal[target]] = [arr[current], arr[target]];
      [valToIdx[arr[current] as number], valToIdx[arr[target] as number]] = [
        current,
        target,
      ];
    }
  }

  // Return swaps
  return swaps;
};

File: minimumSwaps.py
# Source: https://www.hackerrank.com/challenges/minimum-swaps-2/problem
def minimumSwaps(arr):
    swaps = 0
    d1, d2 = {}, {}

    for i in range(len(arr)):
        d1[i] = arr[i]
        d2[arr[i]] = i

    for i in range(len(arr)):
        if d1[i] != i + 1:
            # Do a swap
            target = d2[i + 1]
            arr[i], arr[target] = arr[target], arr[i]
            swaps += 1

            # Update dictionaries
            d1[i], d1[target] = arr[i], arr[target]
            d2[arr[i]], d2[arr[target]] = i, target

    # Return swaps
    return swaps


print(minimumSwaps([5, 1, 3, 2, 6, 4, 7]) == 4)
print(minimumSwaps([3, 7, 6, 9, 1, 8, 10, 4, 2, 5]) == 9)
